mlflow/pyspark/ml/__init__.py

```diff
@@ -107,24 +120,124 @@ def _get_estimator_info_tags(estimator):
+
+    yield stage
+    if isinstance(stage, Pipeline):
+        for stage in stage.getStages():
+            yield from _traverse_stage(stage)
+    elif isinstance(stage, OneVsRest):
+        yield from _traverse_stage(stage.getClassifier())
+    elif _is_parameter_search_estimator(stage):
+        yield from _traverse_stage(stage.getEstimator())
+        yield from _traverse_stage(stage.getEvaluator())
+
+
+def _get_uid_to_indexed_name_map(estimator):
+    counter = defaultdict(int)
+    uid_to_classname_and_count = {}
+    for child in _traverse_stage(estimator):
+        class_name = child.__class__.__name__
+        counter[class_name] += 1
+        uid_to_classname_and_count[child.uid] = (class_name, counter[class_name])
+    return {
+        uid: f"{class_name}_{count}" if counter[class_name] > 1 else class_name
+        for uid, (class_name, count) in uid_to_classname_and_count.items()
+    }
+
+
+def _gen_stage_hierarchy_recursively(
+    stage, uid_to_indexed_name_map,
+):
+    from pyspark.ml import Pipeline
+    from pyspark.ml.classification import OneVsRest
+
+    stage_name = uid_to_indexed_name_map[stage.uid]
+
+    if isinstance(stage, Pipeline):
+        sub_stages = []
+        for sub_stage in stage.getStages():
+            sub_hierarchy = _gen_stage_hierarchy_recursively(sub_stage, uid_to_indexed_name_map)
+            sub_stages.append(sub_hierarchy)
+        return {"name": stage_name, "stages": sub_stages}
+    elif isinstance(stage, OneVsRest):
+        classifier_hierarchy = _gen_stage_hierarchy_recursively(
+            stage.getClassifier(), uid_to_indexed_name_map
+        )
+        return {"name": stage_name, "classifier": classifier_hierarchy}
+    elif _is_parameter_search_estimator(stage):
+        evaluator = stage.getEvaluator()
+        tuned_estimator = stage.getEstimator()
+        return {
+            "name": stage_name,
+            "evaluator": _gen_stage_hierarchy_recursively(evaluator, uid_to_indexed_name_map),
+            "tuned_estimator": _gen_stage_hierarchy_recursively(
+                tuned_estimator, uid_to_indexed_name_map
+            ),
+        }
+    else:
+        return {"name": stage_name}
 
-def _get_instance_param_map_recursively(instance, level):
+
+def _gen_estimator_metadata(estimator):
+    """
+    Returns an AutologgingEstimatorMetadata object.
+    The AutologgingEstimatorMetadata object includes:
+     - hierarchy: the hierarchy of the estimator, it will expand
+         pipeline/meta estimator/param tuning estimator
+     - uid_to_indexed_name_map: a map of `uid` -> `name`, each nested instance uid will be
+         mapped to a fixed name. The naming rule is using `{class_name}` if the
+         instance type occurs once, or `{class_name}_{index}` if the instance type occurs
+         multiple times. The index order is in line with depth-first traversing.
+     - param_search_estimators: a list includes all param search estimators in the
+         hierarchy tree.
```

Comment:
Can a pipeline contains multiple param search estimators?

Reply:
It can. It is legal, but i think no user will do that.